modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/SleepTimer.kt

```diff
@@ -31,171 +30,191 @@ class SleepTimer @Inject constructor(
 
-    private var sleepTimeMs: Long? = null
-    private var lastSleepAfterTime: Duration? = null
-    private var lastSleepAfterEndOfChapterTime: Duration? = null
-    private var lastTimeSleepTimeHasFinished: Duration? = null
-    private var lastEpisodeUuidAutomaticEnded: String? = null
+    private var sleepTimerHistory = SleepTimerHistory()
 
-    fun sleepAfter(duration: Duration, onSuccess: () -> Unit) {
-        val sleepAt = System.currentTimeMillis().milliseconds + duration
+    private val _stateFlow: MutableStateFlow<SleepTimerState> = MutableStateFlow(SleepTimerState())
+    val stateFlow: StateFlow<SleepTimerState> = _stateFlow
+
+    fun getState(): SleepTimerState = _stateFlow.value
 
-        if (createAlarm(sleepAt.inWholeMilliseconds)) {
-            lastSleepAfterTime = duration
-            cancelAutomaticSleepOnEpisodeEndRestart()
-            cancelAutomaticSleepOnChapterEndRestart()
-            onSuccess()
+    fun updateSleepTimerStatus(
+        sleepTimeRunning: Boolean,
+        sleepAfterEpisodes: Int = 0,
+        sleepAfterChapters: Int = 0,
+        timeLeft: Duration? = null,
+    ) {
+        _stateFlow.update { currentState ->
+            currentState.copy(
+                isSleepTimerRunning = sleepTimeRunning,
+                numberOfEpisodesLeft = sleepAfterEpisodes,
+                numberOfChaptersLeft = sleepAfterChapters,
+                timeLeft = if (!sleepTimeRunning || sleepAfterEpisodes != 0 || sleepAfterChapters != 0) {
+                    ZERO
+                } else {
+                    timeLeft ?: currentState.timeLeft
+                },
+            )
         }
     }
 
-    fun addExtraTime(minutes: Int) {
-        val currentTimeMs = sleepTimeMs
-        if (currentTimeMs == null || currentTimeMs < 0) {
+    fun sleepAfter(duration: Duration, onSuccess: () -> Unit) {
+        updateSleepTimerStatus(sleepTimeRunning = true, timeLeft = duration)
+
+        cancelAutomaticSleepOnEpisodeEndRestart()
+        cancelAutomaticSleepOnChapterEndRestart()
+
+        sleepTimerHistory = sleepTimerHistory.copy(
+            lastSleepAfterTime = duration,
+            lastTimeSleepTimeHasFinished = System.currentTimeMillis().milliseconds + duration,
+        )
+
+        onSuccess()
+    }
+
+    fun addExtraTime(duration: Duration) {
+        val currentTimeLeft: Duration = getState().timeLeft
+        if (currentTimeLeft < ZERO) {
             return
         }
-        val time = Calendar.getInstance().apply {
-            timeInMillis = currentTimeMs
-            add(Calendar.MINUTE, minutes)
+        val newTimeLeft = currentTimeLeft + duration
+
+        _stateFlow.update { currentState ->
+            currentState.copy(timeLeft = newTimeLeft)
         }
-        LogBuffer.i(TAG, "Added extra time: $minutes")
-        createAlarm(time.timeInMillis)
+
+        LogBuffer.i(TAG, "Added extra time: $newTimeLeft")
     }
 
     fun restartTimerIfIsRunning(onSuccess: () -> Unit): Duration? {
```

Comment:
I simplified this method by removing all the callbacks since we now have the state within sleep timer